Parse description, Update src in <script>, Add nodes to HTML

dmorejon opened the pull request:
This is a large PR, but mostly due to test changes and input/reference file additions. 
The source code now removes the description() and debug() calls in transform.js, returning the first description() string as a title.
That string is then passed to injectScripts.js and inserted as a `<title>` node, following an order of precedence on location. This includes a new insertion process for adding a node within other tags like `<head>` and `<html>`. There is an associated test for each location a `<title>` can go.
The insertion functions insertNode and insertNodeWithinTag try to correctly indent the new node based on the surrounding context indentation.
The setup() call is now only added once to the first non-empty script.

dmorejon committed:
Add <title>, strip description, debug, index.js add setup once

dmorejon committed:
Added additional documentation, updated index.js transformation.

dmorejon committed:
Updated testTransform to test title return value

Changed transform.js to use the first description() call only.

Hexcles commented on index.js:
```diff
@@ -40,12 +40,33 @@ function transformFile(filePath) {
     originalScripts.forEach((script) => {
-      const newScript = transformSourceCodeString(script);
-      transformedScripts.push(newScript);
+      // Handles the <script src="js-test.js"></script> tags
```
This is now getting a bit complicated. Perhaps we should assert the original and transformed scripts have the same length, partly due to the early return.

dmorejon replied:
Can you clarify a bit more? Not sure I understand.

Hexcles replied:
Oh sorry. Here I see for each original script you push a transformed script to the array. The logic was pretty obvious originally, but is getting a bit complicated in this revision -- for example, you could forget to push an empty script before the early return. Therefore, we might benefit from a bit of defensive programming here by checking the lengths of the two arrays after the forEach block.

Hexcles commented on src/injectScripts.js:
```diff
@@ -1,12 +1,18 @@
 const TEST_HARNESS_REPORT = 'testharnessreport.js';
-const JS_TEST = 'js-test.js';
+
+const INDENTATION_REGEX = new RegExp('^\n[ ]*$');
```
Might as well use `\s` to capture all whitespaces including tabs (yeah there are some...)?

dmorejon replied:
Replacing with /^\n\s*$/g instead since the \s character would need to be double escaped in the RegExp string.

Hexcles commented on test/testdata/input/canvas-description-and-role.html:
```diff
@@ -0,0 +1,40 @@
+description("This test makes sure that a canvas with and without fallback content each has the right role and description.")
+
+if (window.testRunner && window.accessibilityController) {
+    testRunner.dumpAsText();
```
This reminds me of something: we can simply drop `testRunner.dumpAsText()`. It's done by `testharnessreport.js` already. Maybe file an issue and address it later?

dmorejon replied:
Sounds good.

Hexcles reviewed (commented):
(half way through)

dmorejon committed:
Indentation Regex Fix

LukeZielinski commented on src/transform.js:
```diff
@@ -140,12 +140,57 @@ function transformShouldBeEqualToSpecific() {
+// Returns an object with the transformed source code and the test title,
+// if one was parsed from description() calls.
 function transformSourceCodeString(sourceCode, addSetup=true) {
+  const transformInfo = {};
```
can you explain somewhere what transformInfo's structure is intended to be?

dmorejon replied:
Sure, I'll add a comment here. It's basically a hack to pass a mutable object to the plugin returned closure so we can get data from inside the transformation code back to the main function scope.

LukeZielinski commented on test/testdata/input/doctype-title.html:
```diff
@@ -0,0 +1,5 @@
+<!DOCTYPE html>
+<script src="../../../resources/js-test.js"></script>
+<script>
+description("Testing to see if title under DOCTYPE in transformed HTML");
```
would probably be good to test a case with multiple descriptions (where one is used for title and others are dropped)
..somewhere, not necessarily in this file.

dmorejon replied:
That's a transform.js feature, which I test once with 2 description calls in multiline_file.js  in testTransform.js. Should we add more tests than that?

LukeZielinski replied:
Yeah, I noticed `multiline_file.js` but then `transformed_multiline_file.js` has no title (cuz it's js?)

dmorejon replied:
injectScripts takes the title as a parameter. As long as transform.js only returns one title, it's impossible for the HTML to contain more than one <title> due to multiple description() calls.

LukeZielinski replied:
Ok, so for multiline_file.js, the expected title is in the test rather than in the reference file -- sorry for missing that.

The case I was thinking of testing was "multiple description() calls are transformed into a single <title> tag where the title is the 1st description()". Which I don't think we're explicitly checking?

LukeZielinski replied:
Discussed offline, tests cover the case I was after.

Hexcles commented on index.js:
```diff
@@ -40,12 +40,33 @@ function transformFile(filePath) {
+      // Use the first title description that appears in the old file.
+      if (title === '') {
+        title = transformation.title;
+      }
```
Let's add a log here if both `title` and `transformation.title` are non-empty.

Hexcles commented on index.js:
```diff
@@ -40,12 +40,33 @@ function transformFile(filePath) {
+    // If the title is still undefined after searching scripts for definitions,
+    // use the filepath.
+    if (title === '') {
+      title = 'testing ' + filePath;
```
Nit: the "testing" prefix doesn't seem to add much value here. Maybe drop it?

Hexcles commented on src/injectScripts.js:
```diff
@@ -41,8 +46,59 @@ function changeSrcPlugin() {
+    // Default: add title to beginning of document
+    addNode(tree, newTitleNode, () => {
+      return true;
     });
```
```suggestion
    addNode(tree, newTitleNode, () => true);
```

Similar for other lambdas above.

Hexcles commented on src/transform.js:
```diff
@@ -140,12 +140,57 @@ function transformShouldBeEqualToSpecific() {
+// Returns a closure that modifies the parameter, transformInfo
+// so that information gathered during traversal can be used in
+// the outer scope.
+function removeDescription(transformInfo) {
```
Nit: perhaps name it `removeDescriptionFactory` to signify its difference from `removeDebug`?

Hexcles commented on src/transform.js:
```diff
@@ -155,7 +200,8 @@ function transformSourceCodeString(sourceCode, addSetup=true) {
   });
-  return output.code;
+
+  return {code: output.code, title: transformInfo.description};
```
For this kind of non-trivial return values, it's best to have some documentation.

dmorejon replied:
How should this documentation be done? Should I just expand on the comment above the function that mentions the return object?

Hexcles replied:
Yup. You could also use a bit of JSDoc here.

Hexcles reviewed (commented):
With more logging, I think it's time to use a structured logging library like https://www.npmjs.com/package/log, which can be done in a follow-up (file an issue to track it).

On that note, it's generally a good idea to file issues for TODOs and link to the issues in code e.g. `TOOD(#123)`.

dmorejon committed:
PR Comments: comments, style changes, renaming

Hexcles commented on package.json:
```diff
@@ -7,6 +7,7 @@
     "@babel/core": "^7.10.4",
     "flags": "^0.1.3",
     "htmlparser2": "^4.1.0",
+    "log": "^6.0.0",
```
Discussed offline and looks like we'll be using the `debug` package instead.

dmorejon committed:
PR Comments: documentation, changed logging.

Using npm debug package for logging now, will need to change to
configure the enable/disable of logging in different modules.

dmorejon committed:
ESLint fixes

Hexcles commented on src/injectScripts.js:
```diff
@@ -41,21 +46,73 @@ function changeSrcPlugin() {
+    });
+    // <title> exists, don't add a new one.
+    if (foundTitle) {
+      console.log('WARNING: existing <title>, not using description parameter');
```
Use `debug` here?

Hexcles commented on src/injectScripts.js:
```diff
@@ -41,21 +46,73 @@ function changeSrcPlugin() {
+    if (addNodeWithinTag(tree, newTitleNode, (node) => node.tag === HTML)) {
+      return;
+    }
+    // Within <!DOCTYPE html> tag
```
```suggestion
    // After <!DOCTYPE html> tag
```

Hexcles reviewed (approved):
LGTM % nits

dmorejon committed:
Minor PR comments
